gcs/constraints/circuit/circuit.cc

```diff
@@ -0,0 +1,288 @@
+
+        if (current_val == nullopt || cycle_length == length) break;
+
+        proof_step << pos_var_data.at(last_val->raw_value).plus_one_lines.at(current_val->raw_value).geq_line
+                   << " + ";
+        cycle_length++;
+    }
+
+    if (prevent_idx.has_value()) {
+        proof.emit_proof_comment("Preventing sub-cycle for succ[" + to_string(prevent_idx->raw_value) + "] = " + to_string(prevent_value->raw_value));
+        proof_step << pos_var_data.at(prevent_idx->raw_value).plus_one_lines.at(prevent_value->raw_value).geq_line
+                   << " + ";
+    }
+    else {
+        proof.emit_proof_comment("Contradicting sub-cycle");
+    }
+
+    proof.emit_proof_line(proof_step.str(), ProofLevel::Current);
+}
+
+auto gcs::prevent_small_cycles(
+    const vector<IntegerVariableID> & succ,
+    const PosVarDataMap & pos_var_data,
+    const ConstraintStateHandle & unassigned_handle,
+    State & state) -> Inference
+{
+    auto result = Inference::NoChange;
+    auto & unassigned = any_cast<list<IntegerVariableID> &>(state.get_constraint_state(unassigned_handle));
+    auto n = succ.size();
+    auto end = vector<long>(n, -1);
+    auto known_ends = vector<long>{};
+    auto chain_lengths = vector<long>{};
+
+    for (auto var : unassigned) {
+        state.for_each_value_while(var, [&](Integer val) -> bool {
+            auto j0 = val.raw_value;
+            auto length = 0;
+            if (state.has_single_value(succ[j0]) && (end[j0] < 0)) {
+                auto j = j0;
+                do {
+                    j = state.optional_single_value(succ[j])->raw_value;
+                    length++;
+                    // Need to check this in case alldiff hasn't been fully propagated.
+                    if (j == j0) {
+                        state.infer_true(JustifyExplicitly{[&](Proof & proof) {
+                            output_cycle_to_proof(succ, j0, length, pos_var_data, state, proof);
+                        }});
+                        increase_inference_to(result, Inference::Contradiction);
+                        return false;
+                    }
+                } while (state.has_single_value(succ[j]));
+                end[j0] = j;
+                known_ends.emplace_back(j0);
+                chain_lengths.emplace_back(length);
+            }
+
+            return true;
+        });
+
+        if (result == Inference::Contradiction)
+            return result;
+    }
+
+    while (! known_ends.empty()) {
+        auto i = known_ends.back();
+        known_ends.pop_back();
+        auto length = chain_lengths.back();
+        chain_lengths.pop_back();
+        if (cmp_less(length, succ.size() - 1)) {
+            increase_inference_to(result,
+                state.infer(succ[end[i]] != Integer{i}, JustifyExplicitly{[&](Proof & proof) {
+                    output_cycle_to_proof(succ, i, length, pos_var_data, state, proof, Integer{end[i]}, Integer{i});
+                }}));
```

Comment:
Should probably check for contradiction and return immediately after each inference. The API's a bit icky here, but after you do something that infers contradiction, State might end up in an invalid state. Perhaps this should be an exception, which is how some other solvers do it, except that's slow.